tensorflow/core/grappler/optimizers/auto_parallel.cc

```diff
@@ -166,7 +170,15 @@ Status AutoParallel::Initialize(const GrapplerItem& item) {
   std::set<string> dont_replicate_nodes;
   for (const auto& variable : item.MainVariables()) {
     dont_replicate_nodes.insert(variable->name());
+    VariableDef var = variables_[strings::StrCat(variable->name(), ":0")];
+    if (!var.initializer_name().empty()) {
+      dont_replicate_nodes.insert(NodeName(var.initializer_name()));
+    }
+    if (!var.snapshot_name().empty()) {
+      dont_replicate_nodes.insert(NodeName(var.initializer_name()));
```

Comment:
You probably meant to at the snapshot_name to the list of nodes to avoid replicating here.

Snapshots are used as a way to prefetch the content of a variable onto a device. If you start creating multiple replicas, you probably want to prefetch the variables onto the devices where each replica is going to be executed. So you probably should replicate the variable snapshots.